### ISSUE TITLE
2.5.4 consistently fails to finish phase 2 of the backup (protocol 2)

### ISSUE THREAD

**Issue by antymat** [grke/burp#892]
I am trying to do the backup with the 2.5.4 burp on devuan chimera. The burp has been build from sources; the server and client are the same machine. In client configuration the protocol has been set to 2. The backup is started from the command line, with `burp -a b`.
The console output is:
```
BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB 198824384
BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB2021-09-30 01:30:07 +0200: burp[19153] main socket 5: Peer closed SSL session
2021-09-30 01:30:07 +0200: burp[19153] error in backup_phase2_client_protocol2
 198824428

--------------------------------------------------------------------------------
Start time: 2021-09-29 09:57:40
  End time: 2021-09-30 01:30:07
Time taken: 15:32:27
                         New   Changed Duplicate   Deleted     Total |  Scanned
                   ------------------------------------------------------------
             Files:   314085         0         0         0    314085 |  3873997
         Meta data:        1         0         0         0         1 |        1
       Directories:     8851         0         0         0      8851 |   388134
        Hard links:       16         0         0         0        16 |    81746
        Soft links:     3367         0         0         0      3367 |   157054
     Special files:        6         0         0         0         6 |      383
            Blocks:198498102         0 591416346         0 789914448 |        0
       Grand total:198824428         0 591416346         0 790240774 |  4501315
                   ------------------------------------------------------------

             Messages:             0
             Warnings:             0

      Bytes estimated:   2424754731916 (2.21 TB)
      Bytes in backup:             0
       Bytes received:    2324268773 (2.16 GB)
           Bytes sent:   1376945338001 (1.25 TB)
--------------------------------------------------------------------------------
2021-09-30 01:30:07 +0200: burp[19153] Error in backup
2021-09-30 01:30:07 +0200: burp[19153] End backup
```
I run the monitor (`burp -a s`) constantly in screen, but when the backup fails, the monitor fails as well, with the signal 11.

**Comment by grke** [comment]
Hello,
Does the server-side log in the backup working directory say anything useful?

**Comment by antymat** [comment]
Hi!
Unfortunately, the backup directory inside of global/clients/<client>/ is not created; actually, after a failed backup, this directory only contains 2 hidden files (`.command` and `.created`). There is no log(.gz) file at all. I tried with both `working_dir_recovery_method=resume` and `working_dir_recovery_method=delete`.
During backup, the `working` directory is created, and the log file is there, but a the beginning it contains some trivia. I will see what happens there when the backup eventually fails, and report back.

**Comment by antymat** [comment]
Hi.
It just fails. This time the monitor itself did not fail, but the backup did fail, and earlier than usual. The client directory is back to the 2 hidden files described before. syslog does not show anything useful either.

**Comment by antymat** [comment]
I had some fun with SSDs, so I could get back to the issue only recently. The issue is not limited to just one client. It happens for big datasets, though - both failures happened for more than 1TB of data. 
I turned on the failure notification - it apparently was failing on one file, but there is nothing special about it - it is big-ish (255GB), but that's about it. I touched it - no change. I can read it with no issues. 
I have experimented with the "working_dir_recovery_method", but it does not change a thing. 
The client process seems to be just sitting there, doing nothing, and after the "network_timeout" it says, that the backup has failed, and that's it. 
The server process does not seem to be doing anything either. It keeps a "global/data/0000/000F/EA5F.lock" file open, but there seem to be no I/O activity. There is another process on the server that tries to write to "global/clients/<client>/lockfile", but that one does not seem to be doing anything either. 

What else can I check?

**Comment by grke** [comment]
Hello,

I would say that a 255GB file is an extremely large file, but that should be OK 

I wonder if perhaps it is a sparse file containing a lot of null data?

**Comment by grke** [comment]
Sorry, I had a keyboard problem and accidentally closed the issue.

**Comment by grke** [comment]
I imagine that there is something about the file that causes the client to not send any data for some time, which then exceeds the network timeout and the backup gets interrupted.

I'm thinking that maybe it is just reading through a lot of null data and finding nothing to send.

When I get a chance, I will try to create an extremely large file like that and see if I can reproduce it.

**Comment by antymat** [comment]
I looked at the lsof results for the client - all the open files I can see are the sockets and libraries. That baffles me a little - to me it looks like the server is waiting for the client, and the client for the server, and they just sit there. 
Sorry, I got the prefix wrong - it was 255MB, not 255GB file - an iso image (debian-testing-ia64-netinst.iso). There are some stretches of zeroes in there, but not for 4h of munching I think. Would 255GB of zeroes be enough? 
Before I investigate further, I have a particularly slow backup to make, that will take a week to transfer data; after that I can play with the big file of zeroes and gdb, and see, where it's hanging.

**Comment by antymat** [comment]
OK, so the client sits in the select in async_io:155, occasionally venturing back to backup_phase2_client_protocol2:349, but quickly coming back to async_io, as all the tests are negative, and the loop continues. Nothing else happens in the client. 
At the server the situation is the same: it stays in the select in async_io:155, then goes back to the do_backup_phase2_server_protocol2:997 in an endless loop. 

They just wait for each other, not doing anything else.

**Comment by antymat** [comment]
ping

**Comment by antymat** [comment]
Sorry to say that, but it looks to me like you have abandoned this project - the commits are rare, you seem not to have time to fix or investigate bugs. Is that the case?

**Comment by grke** [comment]
Hello,

I agree that I haven't been doing much on burp recently, for various reasons to do with my time being needed elsewhere, but I wouldn't say that the project is abandoned just yet. I've worked on burp for 10 years now, so I will keep going.

I fully understand that from your perspective that's not much help when I go quiet like this.

**Comment by antymat** [comment]
It is a bit frustrating, I admit. 
I have an issue I can reproduce reliably, you know the code. I **can** try and figure out, how the code works, but it would take me a lot of time. You could try to reproduce the bug, but it would take a while, too. That is why I think that it is more productive to join forces to solve this.

**Comment by antymat** [comment]
It's been almost 3 months, there is no meaningful activity, the issue remains unresolved and it starts to affect me. I understand you have no time to work on your project ATM, but I cannot see, and you didn't state, any clear way you want to move forward. 

It's time for me to move on. This project **is** dead. And so is my support.

**Comment by pagalba-com** [comment]
Hi, I noticed your issue, but didn't really linked it to very same problem of mine, as it was on Macos. 
Please see https://github.com/grke/burp/issues/902
I think this can be OOM issue. This could also help me to understand, what are max children setting in your config?
Also how much memory server has?

**Comment by antymat** [comment]
Hi! Thank you for message.
I have the max_children set to 25, there is 32GiB of RAM in my server + 64GiB of swap. There is very little else running there. I doubt memory is the problem.
I have noticed that the backup would fail on exclusions: if the client encounters a file on the exclusion list, the backup would fail. I have removed all the exclusions, it works much better. Still fails in unpredictable ways, though.  And the test can take anywhere from a week to 2 months.
